stage #0/module #4 Algorithms, Data Structures/algorithms_and_data_structures.md

```diff
@@ -0,0 +1,487 @@
+arr[0] = new int[]{1};       // initializes the first row
+arr[1] = new int[]{1, 2};    // initializes the second row
+arr[2] = new int[]{1, 2, 3}; // initializes the third row
+```
+![Ragged two-dimensional array №2](img/ragged_two_dimensional_array_2.png)
+
+### Linked List
+#### Overview
+A linked list is a linear data structure, which that includes a sequence of connected nodes: each node has the data and the address (reference/link) of the next node in the linked list. Such an address is called a "next pointer". Linked lists have a special member called "Head", which gives the address of the first node. If a linked list is empty, a value of "head" will point at "nowhere", e.g. "null". The last node is always points at "null" (as we already discussed at "nowhere"). In other works, a linked list is a bunch of randomly organized memory addresses, which are linked together. Linked list is the second most-used data structure after an array.
+
+![Linked List Overview](img/linked_list_overview.png)
+
+### Need of Linked Lists
+With linked lists, you never have to move your items: linked lists have the ability to break the chain and rejoin it. In other words, if you wanted to put a new element between Node №1 and Node №2, the steps would be:
+- Create a new Node and allocate memory to it;
+
+![Linked List Need №1](img/linked_list_need_1.png)
+
+- Add a new "data" value;
+
+![Linked List Need №2](img/linked_list_need_2.png)
+
+- Point its "next" pointer to Node №2;
+
+![Linked List Need №3](img/linked_list_need_3.png)
+
+- Change the "next" pointer of Node №1 to a newly created Node.
+
+![Linked List Need №4](img/linked_list_need_4.png)
+
+Doing something similar in an array would have required shifting the positions of all the subsequent elements.
+
+### Pros and Cons of Arrays and Linked Lists
+Which data structure should we use: arrays or linked lists? It depends only on a particular use case and goals, which we intend to resolve. In order to make a balanced decision, we should know pros and cons of these data structures.
+
+#### Situation №1: Read of the last element
+A read operation is heavily tied to type of access. There are two types of access to elements of a data structure: random access and sequential access.
+Sequential access means reading all elements one by one starting from the 1st element. Random access means that we can reach a needed element directly. Linked lists can only perform sequential access, when arrays can do both sequential and random types of access.
+
+Reading of the last element may be a quite heavy operation for a linked list: we can't just read it, because you don't know what address it has. We have to go to Node №1 to get the address for Node №2, then until we get to the last Node. Thus, linked lists are great if we are going to read all the Nodes one at a time, and they are terrible, if we are planning to jump around its Nodes: first, access the last element, then the 3rd element, then the last once again.
```

Comment:
things are a bit different, because in implementation linked lists have links to prev objects too, and reading the last element is a very light operation:
![image](https://user-images.githubusercontent.com/56270371/160283928-9a9f4e44-eaa9-4fa8-b45c-bb34031b0f0f.png)
as you can see we are able to get previous node in linked list